pkg/storage/ingest/reader.go

```diff
@@ -900,30 +880,33 @@ func (r *concurrentFetchers) fetchSingle(ctx context.Context, w fetchWant, _ log
 		Topic:   r.topicName,
 		TopicID: r.topicID,
 		Partitions: []kmsg.FetchRequestTopicPartition{{
-			Partition:          r.partitionID,
-			FetchOffset:        w.startOffset,
-			LastFetchedEpoch:   -1,
+			Partition:   r.partitionID,
+			FetchOffset: w.startOffset,
+			// CurrentLeaderEpoch: -1 means we don't know. Kafka brokers are ok with that.
+			// It does mean that we might end up fetching from an out-of-sync replica.
+			// If we provide this the broker would check if we have up-to-date data.
 			CurrentLeaderEpoch: -1,
-			LogStartOffset:     -1,
+			LastFetchedEpoch:   -1,
```

Comment:
this is probably the shoddiest of this implementation. Do we want to deal with this problem? I couldn't find any docs on how to handle this, but I found some implementation notes in franz-go and the list of all possible errors

https://github.com/grafana/mimir/blob/04df05d32320cd19281591a57c19cceba63ceabf/vendor/github.com/twmb/franz-go/pkg/kgo/topics_and_partitions.go#L530-L537

I think we should handle all of them as best as we can. Currently we handle only the most common one.

https://github.com/grafana/mimir/blob/04df05d32320cd19281591a57c19cceba63ceabf/vendor/github.com/twmb/franz-go/pkg/kmsg/generated.go#L4705-L4745

looks like a lot of code we'll have to write and test :( I'll work on this next

Reply:
Ok, I think I get now - this looks like something worth exploring, but we can do it in the next PR. I don't think we need to implement them but just think about what kind of semantics to we want for them and then do it incrementally.

As for the implementation of this request definition itself. Have you tried:

```golang
	req := kmsg.NewFetchRequest()
	req.MinBytes = 1
	req.Version = 13
	req.MaxWaitMillis = 10000
	req.MaxBytes = w.MaxBytes()

	topic := kmsg.NewFetchRequestTopic()
	topic.Topic = r.topicName
	topic.TopicID = r.topicID

	// Using NewFetchRequestTopicPartition gives us a partition with the default values.
	// One of them is CurrentLeaderEpoch: -1, which means we don't know the leader epoch and Kafka brokers are ok with that.
	// It does mean that we might end up fetching from an out-of-sync replica.
	// If we provide this the broker would check if we have up-to-date data.
	partition := kmsg.NewFetchRequestTopicPartition()
	partition.Partition = r.partitionID
	partition.FetchOffset = w.startOffset
	partition.PartitionMaxBytes = req.MaxBytes

	topic.Partitions = append(topic.Partitions, partition)
	req.Topics = append(req.Topics, topic)
```

Upon a closer look:

```golang
CurrentLeaderEpoch: -1,
LastFetchedEpoch:   -1,
LogStartOffset:     -1,
```

The definition of this is already handled in (and potentially more in the future) by:

```golang
func (v *FetchRequestTopicPartition) Default() {
	v.CurrentLeaderEpoch = -1
	v.LastFetchedEpoch = -1
	v.LogStartOffset = -1
}
```

So I would rather use that.

Reply:
> As for the implementation of this request definition itself. Have you tried:

Agree, I totally forgot about these utilities. Changed to use them now